### PR TITLE
Components can shadow rust (wasi) imports with elixir functions

### PR DESCRIPTION
**Pull request by tessi**
Components can now shadow rust (wasi) imports with elixir functions by namespacing imports:

```
%{
  "wasi:random/random@0.2.4" => %{
    # disable randomness by always returnung 0's
    "get-random-bytes" => {:fn, &List.duplicate(0, &1)}
  }
}
```

for example, would shadow the default WASI implementation for `get-random-bytes` (pro tip: look up actual used versions of the package in your generated .wasm file - e.g. using https://www.wa2.io)

In addition, this PR revamps component type conversion:
  * options are now alway either `:none`, or `{:some, term()}` in Elixir
  * results are now always `{:ok, term()}`, `:ok`, `{:error, term()}`, `:error` depending on the WIT type definition
  * enums/variants can use both, atoms or strings, for case-tags
  * implemented missing type conversions for both, exported and imported
  * implemented wasm-component type check tests in rust
  * removed auto-conversion of record fields from/to snake.kebab case as it made it hard to debug record type conversion. let's do any conversion, if needed, in Elixir and stay close to the true thing in Rust
  * wasm-bools can be initialized from any Elixir term (following 'truthy' behaviour of Elixir)
  * implemented flags, charlist support for imported functions
  * removed old type conversion test, and wrote newer and more expansive tests in rust

### CODE REVIEW

**Comment by tessi** [comment]
haha, no fast merge! need to tell CI how to install `cargo component`

File: native/wasmex/src/component_type_conversion.rs [hunk 1]
```diff
@@ -81,53 +95,146 @@ pub fn term_to_val(param_term: &Term, param_type: &Type) -> Result<Val, Error> {
...
                 Ok(Val::Option(None))
             } else {
-                let converted_val = term_to_val(param_term, &option_type.ty())?;
-                Ok(Val::Option(Some(Box::new(converted_val))))
+                Err(Error::Term(Box::new(format!(
+                    "Invalid atom: {}, expected ':none' or '{{:some, term}}' tuple",
+                    none_atom
+                ))))
+            }
+        }
```
Comment (tessi) [inline native/wasmex/src/component_type_conversion.rs:134]: here we need to add a test whether `option_type` is `None` - only then are we allowed to use the `:none` case. 

-> Can be done in a later PR to untangle this change from this already gigantic PR
